blockscout-ens/bens-server/src/settings.rs

```diff
@@ -28,10 +28,49 @@ impl ConfigSettings for Settings {
     const SERVICE_NAME: &'static str = "BENS";
 }
 
+// TODO: move database settings to blockscout-service-launcher
 #[derive(Debug, Clone, Deserialize, PartialEq, Eq)]
 #[serde(deny_unknown_fields)]
 pub struct DatabaseSettings {
-    pub url: String,
+    pub connect: DatabaseConnectSettings,
+}
+
+#[derive(Debug, Clone, Deserialize, PartialEq, Eq)]
+#[serde(deny_unknown_fields, rename_all = "lowercase")]
+pub enum DatabaseConnectSettings {
+    Url(String),
+    KV(DatabaseKVConnection),
+}
+
+impl DatabaseConnectSettings {
+    pub fn url(&self) -> String {
+        match self {
+            DatabaseConnectSettings::Url(s) => s.clone(),
+            DatabaseConnectSettings::KV(kv) => kv.url(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Deserialize, PartialEq, Eq)]
+#[serde(deny_unknown_fields)]
+pub struct DatabaseKVConnection {
+    pub host: String,
+    pub port: i32,
+    pub user: String,
+    pub password: String,
+    #[serde(default)]
+    pub dbname: String,
+    #[serde(default)]
+    pub options: String,
+}
+
+impl DatabaseKVConnection {
+    pub fn url(&self) -> String {
+        format!(
+            "postgresql://{}:{}@{}:{}/{}{}",
```

Comment:
You either have to make `dbname` non-optional (without `default` annotation), or handle the cases when they are missing (excessive `/` symbol in the end)

I suggest to remain it optional but at the same time remove the latest `/` if `dbname` does not exist. The same goes with `options`. It's better to add `?` symbol before the options theselves, and not to ask users to provide such symbol

Reply:
ok